### PR TITLE
Pullbacks of synthetic categories

### PR DESCRIPTION
**Pull request by ivankobe**
This PR contains the axioms for pullbacks of synthetic categories together with the needed infrastructure, such as cospans and cone diagrams.
It seemed impossible to do this with the existing axioms, so I postulated an additional axiom which states that given natural isomorphisms α β γ δ (with appropriate domains and codomains) together with 3-isomorphisms α≅γ and β≅δ, there is an isomorphism of respective horizontal composites: β * α≅δ * γ. I suppose this axiom is validatd by Makkai's principle.
Note also that some lines exceed the maximum character limit. I wasn't sure how to avoid this since I tried to avoid ad-hoc renamings and `let` statement.

### CODE REVIEW

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    (f : functor-Synthetic-Category-Theory κ C D)
+    (g : functor-Synthetic-Category-Theory κ D C) → UU l
+  is-inverse-Synthetic-Category-Theory κ μ ι f g =
+    ( is-section-Synthetic-Category-Theory κ μ ι f g)
+      ×
+    ( is-retraction-Synthetic-Category-Theory κ μ ι f g)
+```
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:42]:
````suggestion
### The predicate of being an inverse of a functor in a synthetic category

```agda
module _
  {l : Level}
  where

  is-inverse-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    (μ : composition-Synthetic-Category-Theory κ)
    (ι : identity-Synthetic-Category-Theory κ)
    {C D : category-Synthetic-Category-Theory κ}
    (f : functor-Synthetic-Category-Theory κ C D)
    (g : functor-Synthetic-Category-Theory κ D C) → UU l
  is-inverse-Synthetic-Category-Theory κ μ ι f g =
    ( is-section-Synthetic-Category-Theory κ μ ι f g)
      ×
    ( is-retraction-Synthetic-Category-Theory κ μ ι f g)
```
````

File: src/synthetic-category-theory/cospans-synthetic-categories.lagda.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+</details>
+
+## Idea
+
+A cospan is a pair of functors f, g with a common codomain:
+
+```text
+C --f--> E <--g-- D
+```
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/cospans-synthetic-categories.lagda.md:29]: This section should mention that we are working with functors of synthetic categories. For example:

````suggestion
A {{#concept "cospan" Disambiguation="Synthetic categories"}} of [synthetic categories](synthetic-category-theory.synthetic-categories.md) is a pair of functors f, g of synthetic categories with a common codomain:

```text
C --f--> E <--g-- D.
```
````

File: src/synthetic-category-theory/cospans-synthetic-categories.lagda.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+module _
+  {l : Level}
+  where
+
+  cospan-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (C E D : category-Synthetic-Category-Theory κ) → UU l
+  cospan-Synthetic-Category-Theory κ C E D =
+    Σ ( functor-Synthetic-Category-Theory κ C E)
+      λ f → functor-Synthetic-Category-Theory κ D E
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/cospans-synthetic-categories.lagda.md:43]: Our convention is to put a pair of parentheses around the last lambda, even if the syntax of Agda technically allows us to omit them. We think it impairs the readability to omit them.

```suggestion
      ( λ f → functor-Synthetic-Category-Theory κ D E)
```

File: src/synthetic-category-theory/cospans-synthetic-categories.lagda.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    (κ : language-Synthetic-Category-Theory l)
+    {C E D : category-Synthetic-Category-Theory κ}
+    (S : cospan-Synthetic-Category-Theory κ C E D) →
+    functor-Synthetic-Category-Theory κ
+      ( right-source-cospan-Synthetic-Category-Theory κ S)
+      ( target-cospan-Synthetic-Category-Theory κ S)
+  right-map-cospan-Synthetic-Category-Theory κ = pr2
+```
+
+## Transformations of cone diagrams
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/cospans-synthetic-categories.lagda.md:89]:
```suggestion
## Transformations of cospan diagrams
```

File: src/synthetic-category-theory/cospans-synthetic-categories.lagda.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+              ( left-map-cospan-Synthetic-Category-Theory κ S)
+              ( χ)
+              ( φ)
+              ( left-map-cospan-Synthetic-Category-Theory κ S'))
+            λ τ →
+              commuting-square-functors-Synthetic-Category-Theory κ μ
+                ( right-map-cospan-Synthetic-Category-Theory κ S)
+                ( χ)
+                ( ψ)
+                ( right-map-cospan-Synthetic-Category-Theory κ S')
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/cospans-synthetic-categories.lagda.md:137]: This was pretty good, but to make it exactly align with our coding conventions, it is better to indent the lambdas two spaces (indenting everything in its scope 2 further spaces too). 

Note that in the following formatting, the two arguments of each Sigma appear exactly below each other and nothing else. This makes it easier to spot how many arguments a function takes, and where those arguments are in the code.

```suggestion
  transformation-cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    (μ : composition-Synthetic-Category-Theory κ)
    {C C' E E' D D' : category-Synthetic-Category-Theory κ}
    (S : cospan-Synthetic-Category-Theory κ C E D)
    (S' : cospan-Synthetic-Category-Theory κ C' E' D') → UU l
  transformation-cospan-Synthetic-Category-Theory κ μ S S' =
    Σ ( functor-Synthetic-Category-Theory κ
        ( left-source-cospan-Synthetic-Category-Theory κ S)
        ( left-source-cospan-Synthetic-Category-Theory κ S'))
      ( λ φ →
        Σ ( functor-Synthetic-Category-Theory κ
            ( target-cospan-Synthetic-Category-Theory κ S)
            ( target-cospan-Synthetic-Category-Theory κ S'))
          ( λ χ →
            Σ ( functor-Synthetic-Category-Theory κ
                ( right-source-cospan-Synthetic-Category-Theory κ S)
                ( right-source-cospan-Synthetic-Category-Theory κ S'))
              ( λ ψ →
                Σ ( commuting-square-functors-Synthetic-Category-Theory κ μ
                    ( left-map-cospan-Synthetic-Category-Theory κ S)
                    ( χ)
                    ( φ)
                    ( left-map-cospan-Synthetic-Category-Theory κ S'))
                  ( λ τ →
                    commuting-square-functors-Synthetic-Category-Theory κ μ
                      ( right-map-cospan-Synthetic-Category-Theory κ S)
                      ( χ)
                      ( ψ)
                      ( right-map-cospan-Synthetic-Category-Theory κ S')))))
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+  is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    (f : functor-Synthetic-Category-Theory κ C D) → UU l
+  is-equiv-Synthetic-Category-Theory κ μ ι f =
+    Σ ( functor-Synthetic-Category-Theory κ _ _)
+      λ g →
+        ( is-inverse-Synthetic-Category-Theory κ μ ι f g)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:69]:
```suggestion
    Σ ( functor-Synthetic-Category-Theory κ _ _)
      ( is-inverse-Synthetic-Category-Theory κ μ ι f)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+  is-inverse-map-inverse-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D}
+    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
+      is-inverse-Synthetic-Category-Theory κ μ ι f
+        ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:93]:
```suggestion
  is-inverse-map-inverse-is-equiv-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    (μ : composition-Synthetic-Category-Theory κ)
    (ι : identity-Synthetic-Category-Theory κ)
    {C D : category-Synthetic-Category-Theory κ}
    {f : functor-Synthetic-Category-Theory κ C D}
    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
    is-inverse-Synthetic-Category-Theory κ μ ι f
      ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+  is-section-map-inverse-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D}
+    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
+      is-section-Synthetic-Category-Theory κ μ ι f
+        ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:104]: We put all the arguments of an infix operation at the same indentation level. In general the rule is: for any function (whether infix or prefix), the arguments are indented the same way (and anything else within the scope of that function is indented deeper).

```suggestion
    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
    is-section-Synthetic-Category-Theory κ μ ι f
      ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+
+  is-retraction-map-inverse-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D}
+    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
+      is-retraction-Synthetic-Category-Theory κ μ ι f
+        ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:116]:
```suggestion
    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
    is-retraction-Synthetic-Category-Theory κ μ ι f
      ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+  is-section-inverse-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D}
+    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
+      is-section-Synthetic-Category-Theory κ μ ι f
+        ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:128]:
```suggestion
    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
    is-section-Synthetic-Category-Theory κ μ ι f
      ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+
+  is-retraction-inverse-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D}
+    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
+      is-retraction-Synthetic-Category-Theory κ μ ι f
+        ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:140]:
```suggestion
    (H : is-equiv-Synthetic-Category-Theory κ μ ι f) →
    is-retraction-Synthetic-Category-Theory κ μ ι f
      ( inverse-is-equiv-Synthetic-Category-Theory κ μ ι H)
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+    is-retraction-map-inverse-is-equiv-Synthetic-Category-Theory κ μ ι H
+
+  section-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D} →
+    is-equiv-Synthetic-Category-Theory κ μ ι f →
+      section-Synthetic-Category-Theory κ μ ι f
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:151]:
```suggestion
    is-equiv-Synthetic-Category-Theory κ μ ι f →
    section-Synthetic-Category-Theory κ μ ι f
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+    is-section-map-inverse-is-equiv-Synthetic-Category-Theory κ μ ι H
+
+  retraction-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D} →
+    is-equiv-Synthetic-Category-Theory κ μ ι f →
+      retraction-Synthetic-Category-Theory κ μ ι f
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:163]:
```suggestion
    is-equiv-Synthetic-Category-Theory κ μ ι f →
    retraction-Synthetic-Category-Theory κ μ ι f
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 14]
```diff
@@ -0,0 +1,315 @@
...
+  where
+
+  is-biinv-is-equiv-Synthetic-Category-Theory :
+    (κ : language-Synthetic-Category-Theory l)
+    (μ : composition-Synthetic-Category-Theory κ)
+    (ι : identity-Synthetic-Category-Theory κ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D} →
+    is-equiv-Synthetic-Category-Theory κ μ ι f →
+      is-biinv-Synthetic-Category-Theory κ μ ι f
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:183]:
```suggestion
    is-equiv-Synthetic-Category-Theory κ μ ι f →
    is-biinv-Synthetic-Category-Theory κ μ ι f
```

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+    (ν : inverse-Synthetic-Category-Theory κ μ ι)
+    (Λ : left-unit-law-composition-Synthetic-Category-Theory κ ι μ)
+    (Ρ : right-unit-law-composition-Synthetic-Category-Theory κ ι μ)
+    (X : horizontal-composition-Synthetic-Category-Theory κ μ)
+    (Α : associative-composition-Synthetic-Category-Theory κ μ)
+    {C D : category-Synthetic-Category-Theory κ}
+    {f : functor-Synthetic-Category-Theory κ C D} →
+    is-biinv-Synthetic-Category-Theory κ μ ι f →
+      is-equiv-Synthetic-Category-Theory κ μ ι f
+  is-equiv-is-biinv-Synthetic-Category-Theory κ μ ι ν Λ Ρ Χ Α B =
```
Comment (EgbertRijke) [inline src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:201]: This definition is fine as it is, but generally I prefer making definitions with copattern matching. To make a definition with copattern matching, press `C-c C-c` within the goal when you're defining it.

**Review by EgbertRijke — COMMENTED** [review]
This PR is getting along nicely! I left mostly remarks about the conventions, and suggestions for some name changes and naming schemes.

File: src/synthetic-category-theory/biinvertible-maps-synthetic-categories.lagda.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+```
+
+### The type of biinvertible functors
+
+```agda
+module _
+  {l : Level}
+  where
+
+  biinv-Synthetic-Category-Theory :
```
[the remark this answers is not in the document]
Comment (ivankobe) [reply src/synthetic-category-theory/biinvertible-maps-synthetic-categories.lagda.md:84]: You're right, `biinv` just sounds kind of awkward.

File: src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md (hunk 15 above)
Comment (ivankobe) [reply src/synthetic-category-theory/equivalences-synthetic-categories.lagda.md:201]: I see,  copattern matching definitely improves readability.

**Comment by EgbertRijke** [comment]
Thanks for making all the fixes. I'll merge this now

**Pull request merged**